File: backend/app/translation/service.py
from __future__ import annotations

import hashlib
from typing import AsyncIterator

from app.config import Settings
from app.observability.metrics import MetricsRegistry
from app.translation.ports import LLMStreamEvent, StreamFinal, TranslationCache, TranslationRequest, TranslationLLM
from app.translation.router import TranslationRouter


class NullTranslationCache(TranslationCache):
    async def get(self, key: str):
        return None

    async def set(self, key: str, value, ttl_seconds: int) -> None:  # noqa: ARG002
        return None


class TranslationService:
    def __init__(
        self,
        settings: Settings,
        router: TranslationRouter,
        cache: TranslationCache | None = None,
        metrics: MetricsRegistry | None = None,
    ) -> None:
        self._settings = settings
        self._router = router
        self._cache = cache or NullTranslationCache()
        self._metrics = metrics or MetricsRegistry()

    def build_cache_key(self, req: TranslationRequest, provider: str, model: str) -> str:
        text_hash = hashlib.sha256(req.text.strip().lower().encode("utf-8")).hexdigest()
        tone = req.tone or "default"
        prompt_ver = req.metadata.get("prompt_version", "v1")
        return f"tr:v2:{provider}:{model}:{req.source_lang}:{req.target_lang}:{tone}:{prompt_ver}:{text_hash}"

    async def translate(self, req: TranslationRequest) -> AsyncIterator[LLMStreamEvent]:
        adapter = self._router.pick(req)
        cache_key = self.build_cache_key(req, adapter.provider, adapter.model)
        cached = await self._cache.get(cache_key)
        if cached:
            self._metrics.increment("cache_hit_rate")
            yield StreamFinal(
                text=cached["text"],
                finish_reason="stop",
                cached=True,
                latency_first_token_ms=0,
                latency_total_ms=0,
            )
            return

        final_event: StreamFinal | None = None
        async for event in adapter.translate_stream(req):
            if isinstance(event, StreamFinal):
                final_event = event
            yield event

        if final_event:
            await self._cache.set(
                cache_key,
                {"text": final_event.text, "provider": adapter.provider, "model": adapter.model},
                self._settings.translation_cache_ttl_seconds,
            )

File: go-app/internal/translation/service.go
package translation

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"strings"

	"talk/go-app/internal/config"
	"talk/go-app/internal/observability"
)

type TranslationService struct {
	settings config.Config
	router   *TranslationRouter
	cache    TranslationCache
	metrics  *observability.MetricsRegistry
}

func NewTranslationService(settings config.Config, router *TranslationRouter, cache TranslationCache, metrics *observability.MetricsRegistry) *TranslationService {
	if cache == nil {
		cache = NewMemoryTranslationCache()
	}
	if metrics == nil {
		metrics = observability.NewMetricsRegistry()
	}
	return &TranslationService{settings: settings, router: router, cache: cache, metrics: metrics}
}

func (s *TranslationService) BuildCacheKey(req TranslationRequest, provider, model string) string {
	hash := sha256.Sum256([]byte(strings.TrimSpace(strings.ToLower(req.Text))))
	tone := "default"
	if req.Tone != nil && *req.Tone != "" {
		tone = *req.Tone
	}
	promptVersion := "v1"
	if req.Metadata != nil {
		if v := req.Metadata["prompt_version"]; v != "" {
			promptVersion = v
		}
	}
	return fmt.Sprintf("tr:v2:%s:%s:%s:%s:%s:%s:%s", provider, model, req.SourceLang, req.TargetLang, tone, promptVersion, hex.EncodeToString(hash[:]))
}

func (s *TranslationService) Translate(req TranslationRequest) ([]StreamEvent, error) {
	adapter := s.router.Pick(req)
	cacheKey := s.BuildCacheKey(req, adapter.Provider(), adapter.Model())
	if cached := s.cache.Get(cacheKey); cached != nil {
		s.metrics.Increment("cache_hit_rate", 1)
		text := cached["text"].(string)
		provider := cached["provider"].(string)
		model := cached["model"].(string)
		zero := 0
		return []StreamEvent{StreamStart{Provider: provider, Model: model}, StreamFinal{Text: text, FinishReason: "stop", Cached: true, LatencyFirstTokenMS: &zero, LatencyTotalMS: &zero}}, nil
	}
	events, err := adapter.TranslateStream(req)
	if err != nil {
		return nil, err
	}
	var final *StreamFinal
	for _, event := range events {
		if v, ok := event.(StreamFinal); ok {
			vv := v
			final = &vv
		}
	}
	if final != nil {
		s.metrics.Increment("cache_write_total", 1)
		s.cache.Set(cacheKey, map[string]any{"text": final.Text, "provider": adapter.Provider(), "model": adapter.Model()}, s.settings.TranslationCacheTTLSeconds)
	}
	return events, nil
}
